### ISSUE TITLE
Error when trying to displaying a vocabulary from a Virtuoso SPARQL endpoint

### ISSUE THREAD

**Issue by tfrancart** [NatLibFi/Skosmos#632]
## At which URL did you encounter the problem?

http://vocabs.sparna.fr/skosmos/legal-institution-history-casemates/fr/

## What steps will reproduce the problem?
1. Go to the given URL
2. ... or go to http://vocabs.sparna.fr and click on "Liste historique des institutions - CASEMATES"

## What is the expected output? What do you see instead?

Expected to see the vocabulary page, got a blank page instead.

The vocabulary configuration is

```
:legal-institution-history-casemates a skosmos:Vocabulary, void:Dataset ;
        dc:title "Liste historique des institutions - CASEMATES"@fr ;
        dc:subject :cat_general ;
        dc:type mdrtype:THESAURUS ;
        void:uriSpace "http://data.legilux.public.lu/resource/authority/legal-institution/historique/";
        skosmos:language "fr";
        skosmos:shortName "Liste historique des institutions - CASEMATES";
        skosmos:fullAlphabeticalIndex "true";
        skosmos:showTopConcepts "true";
        void:sparqlEndpoint <http://data.legilux.public.lu/sparql> ;
        skosmos:sparqlGraph <http://data.legilux.public.lu/resource/authority/legal-institution-history/graph> .
```

The Apache log says :

```
[Wed Oct 04 16:58:31.888067 2017] [:error] [pid 19071] [client 109.74.87.2:52432] PHP Fatal error:  Uncaught Error: Call to undefined method EasyRdf_Sparql_Result::resource() in /var/lib/skosmos/model/Vocabulary.php:144\nStack trace:\n#0 /var/lib/skosmos/vendor/twig/twig/lib/Twig/Template.php(680): 
Vocabulary->getInfo('fr')\n#1 /tmp/skosmos-template-cache/b4/b438a52156ba66e79e559db8c29b8e52729c0ad9707262a507b62dae7412cf29.php(33): Twig_Template->getAttribute(Object(Vocabulary), 'info', Array, 'method')\n#2 /var/lib/skosmos/vendor/twig/twig/lib/Twig/Template.php(438): 
__TwigTemplate_582ccf7fc687f49fd0eba2baa31441b1f2ce99184d5fda00ae3ba6d804d3ef3a->doDisplay(Array, Array)\n#3 /var/lib/skosmos/vendor/twig/twig/lib/Twig/Template.php(406): 
Twig_Template->displayWithErrorHandling(Array, Array)\n#4 /tmp/skosmos-template-cache/9e/9e144ce04ac068de9c5f43446bf36669bd75f2b3d8970b749ed17f27e4e2e40f.php(188): 
Twig_Template->display(Array)\n#5 /var/lib/skosmos/vendor/twig/twig/lib/Twig/Template.php(215): 
__TwigTemplate_72284e08262d094ed9ca748c5808d2ecb1cd528dd6e930f9045576b893892b0a->block_content(Array, Array)\n#6 /tmp/skosmos-t in /var/lib/skosmos/model/Vocabulary.php on line 144, referer: http://vocabs.sparna.fr/skosmos/fr/
```

I have updated and reinstalled the dependencies to make use EasyRDF was installed with the correct version.
I have no clue what can possibly go wrong, any idea would be appreciated. The endpoint at http://data.legilux.public.lu/sparql is Virtuoso, could it be a problem with how the results are returned from the endpoint ?

(Note also that the same data, when loaded in a local Fuseki endpoint, works fine; so this is not a data-related issue, it is related to querying a remote endpoint).

Thanks

**Comment by osma** [comment]
To Skosmos, all endpoints are "remote", they are all accessed via HTTP and it makes no difference if they're on localhost or on the opposite side of the world (except for latency).

Maybe you could try enabling the logging functionality in `config.inc`? That would give you an idea of what SPARQL queries Skosmos is trying to perform and perhaps see the response from the endpoint.

It would also be good to know whether this affects just a single query (this seems to be about general information for the vocabulary) or all queries, or something in between. You could try accessing a concept page directly, or the REST API. These would trigger other SPARQL queries than the one that currently fails.

It's possible that this is an incompatibility between EasyRdf and Virtuoso. I remember debugging similar problems before. For example the content negotiation could go wrong. The details are different for CONSTRUCT and SELECT queries, so one type might work while the other fails.

**Comment by tfrancart** [comment]
Accessing a concept page directly : http://vocabs.sparna.fr/skosmos/legal-institution-history-casemates/fr/page/1

Same behavior. Logs :

```
[Tue Oct 10 09:39:13.004893 2017] [:error] [pid 6752] [client 109.74.87.2:37376] PHP Fatal error:  Uncaught Error: Call to undefined method EasyRdf_Sparql_Result::isEmpty() in /var/lib/skosmos/model/sparql/GenericSparql.php:484\nStack trace:\n#0 /var/lib/skosmos/model/Vocabulary.php(391): GenericSparql->queryConceptInfo(Array, NULL, Array, 'fr')\n#1 /var/lib/skosmos/controller/WebController.php(215): Vocabulary->getConceptInfo('http://data.leg...', 'fr')\n#2 /var/lib/skosmos/index.php(87): WebController->invokeVocabularyConcept(Object(Request))\n#3 {main}\n  thrown in /var/lib/skosmos/model/sparql/GenericSparql.php on line 484
```

Calling a REST API works : http://vocabs.sparna.fr/skosmos/rest/v1/legal-institution-history-casemates/vocabularyStatistics?lang=fr. Or http://vocabs.sparna.fr/skosmos/rest/v1/legal-institution-history-casemates/topConcepts?lang=fr.
So as you say it could be with some SPARQL queries only. Typically something related to request or response content type ? (I had a similar problem with unexpected content types for ASK queries in Virtuoso. SPARQL does not define formally the response content type for ASK queries. I fixed this by adjusting the Accept header of the HTTP request to request "application/json" which garantees the content type returned).

**Comment by tfrancart** [comment]
Additional question to refine my analysis : how can I log the SPARQL queries being sent by Skosmos ? I see there is a query logger mechanism, but I don't know how to activate it and where the logs end  up ?

**Comment by osma** [comment]
@tfrancart simply set in config.inc:
```
define ("LOG_FILE_NAME", "/path/to/logfile");
```

**Comment by tfrancart** [comment]
Thanks, I should have seen this. Here is the latest quesry I see in the log file :

```
[2017-10-10 09:58:20] INFO [qid 99385] SPARQL query:
CONSTRUCT {
  <http://data.legilux.public.lu/resource/authority/legal-institution-history> ?property ?value .
} FROM <http://data.legilux.public.lu/resource/authority/legal-institution-history/graph> WHERE {
  <http://data.legilux.public.lu/resource/authority/legal-institution-history> ?property ?value .
  FILTER (?property != skos:hasTopConcept)
}
[2017-10-10 09:58:20] INFO [qid 99385] result: 12 rows returned in 106 ms
```

It seems it is executed correctly. Then the error in the apache log :

```
Tue Oct 10 09:58:20.240189 2017] [:error] [pid 1314] [client 109.74.87.2:37562] PHP Fatal error:  Uncaught Error: Call to undefined method EasyRdf_Sparql_Result::resource() in /var/lib/skosmos/model/Vocabulary.php:144\nStack trace:\n#0 /var/lib/skosmos/vendor/twig/twig/lib/Twig/Template.php(680): Vocabulary->getInfo('fr')\n#1 /tmp/skosmos-template-cache/b4/b438a52156ba66e79e559db8c29b8e52729c0ad9707262a507b62dae7412cf29.php(33): Twig_Template->getAttribute(Object(Vocabulary), 'info', Array, 'method')\n#2 /var/lib/skosmos/vendor/twig/twig/lib/Twig/Template.php(438): __TwigTemplate_582ccf7fc687f49fd0eba2baa31441b1f2ce99184d5fda00ae3ba6d804d3ef3a->doDisplay(Array, Array)\n#3 /var/lib/skosmos/vendor/twig/twig/lib/Twig/Template.php(406): Twig_Template->displayWithErrorHandling(Array, Array)\n#4 /tmp/skosmos-template-cache/9e/9e144ce04ac068de9c5f43446bf36669bd75f2b3d8970b749ed17f27e4e2e40f.php(188): Twig_Template->display(Array)\n#5 /var/lib/skosmos/vendor/twig/twig/lib/Twig/Template.php(215): __TwigTemplate_72284e08262d094ed9ca748c5808d2ecb1cd528dd6e930f9045576b893892b0a->block_content(Array, Array)\n#6 /tmp/skosmos-t in /var/lib/skosmos/model/Vocabulary.php on line 144, referer: http://vocabs.sparna.fr/skosmos/fr/
```

**Comment by osma** [comment]
(moving the comment here since I also fell for the wrong issue thing...)

I suspect this is a content negotiation problem between EasyRdf and GraphDB. Is there any chance you could peek at the relevant HTTP headers, especially Accept (request) and Content-Type (response)?

You could do this with ngrep watching for traffic, or using a HTTP proxy with logging, or by editing the EasyRdf code (Client.php IIRC)

You could also write a minimal PHP script that runs the CONSTRUCT query using EasyRdf.

If GraphDB returns something that EasyRdf doesn't recognize as an RDF graph, things go wrong...

**Comment by osma** [comment]
My theory:

1. EasyRdf sends a CONSTRUCT query with a rather broad Accept header value
2. GraphDB responds with something, with Content-type X (maybe `application/sparql-results+xml`?)
3. EasyRdf sees the Content-type X and parses the response as a SPARQL result set (not RDF graph)
4. When Skosmos tries to access a resource in the response, an exception is thrown

**Comment by tfrancart** [comment]
OK, as far as I can tell, here are the request and response details. This is a SELECT query :

HTTP query sent by Skosmos :

```
http://data.legilux.public.lu/sparql?query=PREFIX+dc%3A+%3Chttp%3A%2F%2Fpurl.org%2Fdc%2Fterms%2F%3E%0APREFIX+dc11%3A+%3Chttp%3A%2F%2Fpurl.org%2Fdc%2Felements%2F1.1%2F%3E%0APREFIX+rdfs%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0APREFIX+skos%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2004%2F02%2Fskos%2Fcore%23%3E%0ASELECT+%3Fcs+%3Flabel+%3Fpreflabel+%3Ftitle+FROM+%3Chttp%3A%2F%2Fdata.legilux.public.lu%2Fresource%2Fauthority%2Flegal-institution-history%2Fgraph%3E%0AWHERE+%7B%0A+%3Fcs+a+skos%3AConceptScheme+.%0A+OPTIONAL+%7B%0A+++%3Fcs+rdfs%3Alabel+%3Flabel+.%0A+++FILTER%28langMatches%28lang%28%3Flabel%29%2C+%27fr%27%29%29%0A+%7D%0A+OPTIONAL+%7B%0A+++%3Fcs+skos%3AprefLabel+%3Fpreflabel+.%0A+++FILTER%28langMatches%28lang%28%3Fpreflabel%29%2C+%27fr%27%29%29%0A+%7D%0A+OPTIONAL+%7B%0A+++%7B+%3Fcs+dc11%3Atitle+%3Ftitle+%7D%0A+++UNION%0A+++%7B+%3Fcs+dc%3Atitle+%3Ftitle+%7D%0A+++FILTER%28langMatches%28lang%28%3Ftitle%29%2C+%27fr%27%29%29%0A+%7D%0A%7D+ORDER+BY+%3Fcs
```

Headers :

```
HTTP/1.1
Connection: close
User-Agent: EasyRdf_Http_Client
Accept: application/sparql-results+json,application/n-triples,application/x-httpd-php-source,application/json,application/ld+json,application/x-ntriples;q=0.9,application/ntriples;q=0.9,text/ntriples;q=0.9,text/plain;q=0.9,application/rdf+json;q=0.9,text/json;q=0.9,application/sparql-results+xml;q=0.8,text/turtle;q=0.8,application/rdf+xml;q=0.8,application/turtle;q=0.7,application/x-turtle;q=0.7,text/html;q=0.4,application/xhtml+xml;q=0.4
```

Result :

```
{
  "head": {
    "link": [],
    "vars": [
      "cs",
      "label",
      "preflabel",
      "title"
    ]
  },
  "results": {
    "distinct": false,
    "ordered": true,
    "bindings": [
      {
        "cs": {
          "type": "uri",
          "value": "http://data.legilux.public.lu/resource/authority/legal-institution-history"
        },
        "preflabel": {
          "type": "literal",
          "xml:lang": "fr",
          "value": "Liste historique des institutions"
        },
        "title": {
          "type": "literal",
          "xml:lang": "fr",
          "value": "Liste historique des institutions"
        }
      }
    ]
  }
}
```

With headers :

```
Accept-Ranges →bytes
Connection →Keep-Alive
Content-Length →460
Content-Type →application/sparql-results+json
Date →Thu, 19 Oct 2017 09:09:30 GMT
Keep-Alive →timeout=5, max=100
Server →Virtuoso/07.20.3219 (Linux) x64_64-generic-linux-glibc2.17-64 VDB
```

And Apache Error :

```
[Thu Oct 19 11:13:58.292496 2017] [:error] [pid 4186] [client 127.0.0.1:34962] PHP Fatal error:  Uncaught Error: Call to undefined method EasyRdf_Sparql_Result::resource() in /var/www/html/skosmos-1.9/model/Vocabulary.php:144\nStack trace:\n#0 /var/www/html/skosmos-1.9/vendor/twig/twig/lib/Twig/Template.php(680): Vocabulary->getInfo('fr')\n#1 /tmp/skosmos-template-cache/b4/b438a52156ba66e79e559db8c29b8e52729c0ad9707262a507b62dae7412cf29.php(33): Twig_Template->getAttribute(Object(Vocabulary), 'info', Array, 'method')\n#2 /var/www/html/skosmos-1.9/vendor/twig/twig/lib/Twig/Template.php(438): __TwigTemplate_582ccf7fc687f49fd0eba2baa31441b1f2ce99184d5fda00ae3ba6d804d3ef3a->doDisplay(Array, Array)\n#3 /var/www/html/skosmos-1.9/vendor/twig/twig/lib/Twig/Template.php(406): Twig_Template->displayWithErrorHandling(Array, Array)\n#4 /tmp/skosmos-template-cache/9e/9e144ce04ac068de9c5f43446bf36669bd75f2b3d8970b749ed17f27e4e2e40f.php(188): Twig_Template->display(Array)\n#5 /var/www/html/skosmos-1.9/vendor/twig/twig/lib/Twig/Template.php(215): __TwigTemplate_72284e08262d094ed9ca748c5808d2ecb1cd528dd6e930f9045576b893892b0a-> in /var/www/html/skosmos-1.9/model/Vocabulary.php on line 144, referer: http://localhost/skosmos-1.9/fr/
```

**Comment by osma** [comment]
Thanks for the detailed debugging. I'm scratching my head at this. I don't know how the Skosmos code is supposed to work! It seems to me that Skosmos is sending a SELECT query but then expecting a graph result, since it's calling the `resource` method which doesn't exist in the `EasyRdf_Sparql_Result` class that is used for SELECT results.

**Comment by osma** [comment]
Are you sure you're using unmodified Skosmos code?

[Vocabulary::getInfo](https://github.com/NatLibFi/Skosmos/blob/master/model/Vocabulary.php#L143) calls [GenericSparql::queryConceptScheme](https://github.com/NatLibFi/Skosmos/blob/master/model/sparql/GenericSparql.php#L585) which then calls [generateConceptSchemeQuery](https://github.com/NatLibFi/Skosmos/blob/master/model/sparql/GenericSparql.php#L567) to generate the query, then executes it and returns the result. This is a CONSTRUCT query, so the result is an `EasyRdf_Graph` and then Vocabulary::getInfo calls `resource` on it.

But the query you show is a SELECT query. It is generated by a very similarly named method [GenericSparql::generateConceptSchemesQuery](https://github.com/NatLibFi/Skosmos/blob/master/model/sparql/GenericSparql.php#L595) (note the plural!). Since the result is not a graph in this case, calling `resource` on it throws an exception. But this should not happen with the vanilla Skosmos code, that's why I'm wondering if you've made some changes and not noticed the difference between the two very similarly named methods.

**Comment by osma** [comment]
The other possible explanation is that you captured the wrong SPARQL query. There should be a rather small CONSTRUCT query sent to the SPARQL endpoint, and the error you see happens when processing the result of that query.

**Comment by tfrancart** [comment]
Hello Osma

Thanks for the feedback. The endpoint is down now, so I can't test further, but note that : 
- Yes, I am using an unmodified version of Skosmos on this server
- I haven't set the default concept scheme in the vocabulary definition, so Vocabulary::getDefaultConceptScheme is issuing the SELECT query above to return the list of ConceptSchemes and picking up the first one (maybe the behavior will change if I set the default concept scheme in vocabularies.ttl, I will let you know)
- So I am positive that the SELECT query above is the first one issued by SKOSMOS when accessing http://vocabs.sparna.fr/skosmos/legal-institution-history-casemates/fr/.

Will let you know the results of further tests when Virtuoso SPARQL endpoint is available again.

**Comment by osma** [comment]
OK so it's the first query, but I think more interesting would be the *last* query performed just before the error. I suspect it's a CONSTRUCT query where Virtuoso returns something that EasyRdf doesn't parse as a graph.

**Comment by tfrancart** [comment]
Osma, you are right. The proxy I used to debug the HTTP query sent made the first query failed, while somehow when not using the proxy it succeeds. This led me to suspect that this first query was the cause of the problem, while actually, without using the proxy, I can see from the Skosmos log that the second one (a CONSTRUCT query) is actually the cause :

```
[2017-10-24 14:46:42] INFO [qid 34230] SPARQL query:
CONSTRUCT {
  <http://publications.europa.eu/resource/authority/language> ?property ?value .
}  WHERE {
  <http://publications.europa.eu/resource/authority/language> ?property ?value .
  FILTER (?property != skos:hasTopConcept)
}

[2017-10-24 14:46:42] INFO [qid 34230] result: 7 rows returned in 52 ms
```

Interestingly, you can see from the logs `7 rows returned` that EasyRDF mistaken the result as a Tuple query result instead of a Graph query result (otherwise we would have `7 triples returned`).


I can see that issuing this HTTP query :

```
GET /sparql?query=CONSTRUCT {   <http://publications.europa.eu/resource/authority/language> ?property ?value . }  WHERE {   <http://publications.europa.eu/resource/authority/language> ?property ?value .   FILTER (?property != skos:hasTopConcept) } HTTP/1.1
Host: data.legilux.public.lu
Accept: application/sparql-results+json,application/n-triples,application/x-httpd-php-source,application/json,application/ld+json,application/x-ntriples;q=0.9,application/ntriples;q=0.9,text/ntriples;q=0.9,text/plain;q=0.9,application/rdf+json;q=0.9,text/json;q=0.9,application/sparql-results+xml;q=0.8,text/turtle;q=0.8,application/rdf+xml;q=0.8,application/turtle;q=0.7,application/x-turtle;q=0.7,text/html;q=0.4,application/xhtml+xml;q=0.4
Cache-Control: no-cache
Postman-Token: cb92ef82-031a-3dea-4c0b-9102dc273434
```

Actually returns a SPARQL results in JSON ! :

```
{
  "head": {
    "link": [],
    "vars": [
      "s",
      "p",
      "o"
    ]
  },
  "results": {
    "distinct": false,
    "ordered": true,
    "bindings": [
      {
        "s": {
          "type": "uri",
          "value": "http://publications.europa.eu/resource/authority/language"
        },
        "p": {
          "type": "uri",
          "value": "http://www.w3.org/2000/01/rdf-schema#label"
        },
        "o": {
          "type": "literal",
          "value": "Langues"
        }
      },
      {
        "s": {
          "type": "uri",
          "value": "http://publications.europa.eu/resource/authority/language"
        },
        "p": {
          "type": "uri",
          "value": "http://publications.europa.eu/ontology/authority/table.id"
        },
        "o": {
          "type": "literal",
          "value": "language"
        }
      },
      ...
      }
    ]
  }
}
```

So probably Virtuoso match on the first "Accept" header (`application/sparql-results+json`) and returns its results this way. I suspect this is not conformant to the SPARQL spec (?).

One way to workaround this would be to have the ability to pass additionnal Accept header values to the EasyRDF client (I see some headers are already set in GenericSPARQL::initializeHttpClient). This could be an additionnal entry in Skosmos config file. What do you think ?

**Comment by osma** [comment]
Just as I thought. Virtuoso responds to a CONSTRUCT query with a set of variable bindings instead of a graph. Not sure whether this is conformant to the SPARQL spec or not. On the HTTP level this is correct, because EasyRdf says it will accept `application/sparql-results+json` and even gives it a higher (implicit) weight than for any of the RDF graph serialization MIME types.

What I found in the [SPARQL 1.1 Protocol](https://www.w3.org/TR/2013/REC-sparql11-protocol-20130321/#query-operation) spec is this:

> The response to a query request is either the SPARQL XML Results Format, the SPARQL JSON Results Format, the SPARQL CSV/TSV Results Format, or an RDF serialization, depending on the query form [SPARQL] and content negotiation [RFC2616]. 

But this doesn't really say whether the query form or the content negotiation is more important.

OTOH, the [SPARQL 1.1 Query Language](https://www.w3.org/TR/sparql11-query/#construct) spec is quite clear:
> The CONSTRUCT query form returns a single RDF graph specified by a graph template. The result is an RDF graph [...]

But whether the Virtuoso response with `?s ?p ?o` bindings is an acceptable rendering of an RDF graph, I don't know.

Nevertheless, this is a general compatibility issue between EasyRdf and Virtuoso, not specific to Skosmos. It should be possible to demonstrate this in a very small PHP script that just performs a CONSTRUCT query to a Virtuoso endpoint using EasyRdf. Maybe you could raise it in the [EasyRdf issue tracker](https://github.com/njh/easyrdf/issues)? Though unfortunately the project is not very active nowadays.

Not sure which software is more at fault here. EasyRdf clearly uses an overly broad `Accept:` value, while Virtuoso responds with a non-standard(?) rendering of an RDF graph.

**Comment by tfrancart** [comment]
I see from https://github.com/njh/easyrdf/blob/master/lib/Sparql/Client.php#L339 that :

- the latest version of EasyRdf completely resets the Accept headers before issuing a query, so whatever we set on the HttpClient at initialization time is overriden
- the latest version of EasyRdf includes a test to set proper Accept headers depending on the query type. This is the latest commit on this file (https://github.com/njh/easyrdf/commit/6314d6ae66b6eeae086278e0a4878a9754aceb5e). However this is not included in the latest release of EasyRdf (0.9).

This could solve our problem; I will give it a try, although it is not included in a release of EasyRdf

**Comment by osma** [comment]
Good to know. There hasn't been a formal EasyRdf release for a long time. There are tags for `0.9.1` and `0.10.0-alpha.1`. We could perhaps depend on the latter tag, since it includes the Accept header changes you mention.

**Comment by tfrancart** [comment]
I updated composer.json with :

```
  "repositories": {
    "easyrdf-local": {
        "type": "vcs",
        "url": "https://github.com/njh/easyrdf"
    },
    ...
    "easyrdf/easyrdf": "dev-master",
```

I fear 0.10.0-alpha.1 introduces incompatible changes, by renaming EasyRdf_Namespace into EasyRdf_RdfNamespace.
I tried to search/replace in the entire Skosmos code, but I still get a `Class 'EasyRdf_RdfNamespace' not found`. I am not familiar with Composer, so I don't know if what I did is correct.

Since this may impact the whole behavior of the application, would you be able to create a separate branch and try to upgrade Skosmos to work with EasyRdf master or 0.10.0-alpha.1 ? this would be terrific. I could try to dig deeper, but this might be over my competencies (not being a PHP developer, and not familiar with Composer dependency loading etc.).

**Comment by osma** [comment]
I think you need to run `composer update` a second time so that class autoloading works.

But I will try making a branch as you suggested.

**Comment by osma** [comment]
The problem is really that basically all EasyRdf classes have been renamed. From CHANGELOG.md:

> * Classes are renamed like this: `EasyRdf_Parser_Turtle` → `EasyRdf\Parser\Turtle`. With a single exception: `EasyRdf_Namespace` → `EasyRdf\RdfNamespace` (because `namespace` is a keyword in PHP)

**Comment by osma** [comment]
Created a separate issue for the upgrade (#643) as it is a more general task than fixing this specific issue (which may or may not be resolved by upgrading)

**Comment by osma** [comment]
@tfrancart want to test the branch issue643-easyrdf-upgrade?

**Comment by tfrancart** [comment]
It works fine with the latest EasyRDF ! Thanks again !

**Comment by osma** [comment]
Resolved via #643
